[FIX] Recommend SI units formatting to adhere to CMIXF-12 (#411)
bids-standard/bids-specification: src/99-appendices/05-units.md

# Appendix V: Units

> As described in the [units section](../02-common-principles.md#units),

Reviewer: ```suggestion
As described in the [Units](../02-common-principles.md#units),
```

# Appendix V: Units

Following the International System of Units (SI, abbreviated from the French
Système international (d'unités)). The ASCII representation is taken from
[here](http://people.csail.mit.edu/jaffer/MIXF/CMIXF-12), where additional units
may be found. The primary reason for choosing an ASCII representation, is that
certain characters can have multiple unicode representations that look very
similar. The BIDS specification will allow only the following unicode symbols for

> backwards compatibility: U+00B5 (µ) and U+00B0 (°) and U+2126 (Ω).

Reviewer: I might rewrite this addition:

```
The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention for encoding units
is RECOMMENDED to achieve maximum portability and limited variability of representation.
Earlier versions of the BIDS standard listed the following Unicode symbols, and these are still
included for backwards compatibility: `U+00B5` (µ) and `U+00B0` (°) and `U+2126` (Ω).
It is REQUIRED that units be CMIXF-12 compliant or among these three Unicode characters.
```

# Appendix V: Units

As described in the [units section](../02-common-principles.md#units),
the specification of units MUST follow the
[International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units)
(SI, abbreviated from the French Système international (d'unités)).

The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention

> for encoding units is RECOMMENDED to achieve maximum portability and limited

Author: ```suggestion
for encoding units is REQUIRED to achieve maximum portability and limited
```

> variability of representation.

Author: ```suggestion
variability of representation.
In case that a CMIXF representation of SI units is not possible, you will have
to declare your units as custom units and define them in accompanying JSON
files, as described in the [units section](../02-common-principles.md#units).
```

# Appendix V: Units

As described in the [units section](../02-common-principles.md#units),
the specification of units MUST follow the
[International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units)
(SI, abbreviated from the French Système international (d'unités)).

The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
for encoding units is REQUIRED to achieve maximum portability and limited
variability of representation.

> In case that a CMIXF representation of SI units is not possible, you will have
> to declare your units as custom units and define them in accompanying JSON
> files, as described in the [units section](../02-common-principles.md#units).

Reviewer: Eliminate "you" phrasing:

```suggestion
If a CMIXF-12 representation of a unit is not possible, the unit MUST be declared
as custom units and defined in an accompanying JSON file, as described in the
[units section](../02-common-principles.md#units).
```

Earlier versions of the BIDS standard listed the following Unicode symbols, and
these are still included for backwards compatibility:

1.  `U+00B5` (µ)
1.  `U+00B0` (°)
1.  `U+2126` (Ω).

It is REQUIRED that units be CMIXF-12 compliant or among these three Unicode
characters.

> Please take care to not confuse the accepted unicode `U+00B5` (µ) with the
> unicode NOT accepted `U+03BC` (μ):

Reviewer: Minor grammar issues, and added links to a useful reference that can give additional context.

```suggestion
Please take care not to confuse the accepted Unicode character
[`U+00B5`](https://codepoints.net/U+00B5) (µ) with the disallowed Unicode
character [`U+03BC`](https://codepoints.net/U+03BC) (μ):
```

They look equal in appearance but are different to the computer.

Units MUST consist of the `unit symbol` with an optionally accompanying
`prefix symbol` (see table below). Appropriate upper- or lower- casing MUST
be applied as declared by CMIXF-12.

For cases that are unspecified by this appendix, or the
[units section](../02-common-principles.md#units), the
[CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
applies.
You can use the [cmixf Python package](https://github.com/sensein/cmixf) to
check whether your formatting is compliant.

Examples:

1.  `uV` or `µV` are permissible, but NOT: `microV`, `µvolt`, `1e-6V`, etc.
1.  Combinations of units are allowed, e.g., `V/us` for the [Slew rate](https://en.wikipedia.org/wiki/Slew_rate)

| Unit name      | Unit symbol       | Quantity name                              |
| -------------- | ----------------- | ------------------------------------------ |
| metre          | m                 | length                                     |

> | kilogram       | kg                | mass                                       |

Reviewer: We're missing liters.

```suggestion
| kilogram       | kg                | mass                                       |
| litre (liter)  | L                 | volume                                     |
```

# Appendix V: Units

As described in the [units section](../02-common-principles.md#units),
the specification of units MUST follow the
[International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units)
(SI, abbreviated from the French Système international (d'unités)).

The [CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
for encoding units is REQUIRED to achieve maximum portability and limited
variability of representation.
If a CMIXF-12 representation of a unit is not possible, the unit MUST be declared
as custom units and defined in an accompanying JSON file, as described in the
[units section](../02-common-principles.md#units).
Earlier versions of the BIDS standard listed the following Unicode symbols, and
these are still included for backwards compatibility:

> 1.  `U+00B5` (µ)
> 1.  `U+00B0` (°)
> 1.  `U+2126` (Ω).

Author: ```suggestion
1.  [`U+00B5` (µ)](https://codepoints.net/U+00B5)
1.  [`U+00B0` (°)](https://codepoints.net/U+00B0)
1.  [`U+2126` (Ω)](https://codepoints.net/U+2126)
```

It is REQUIRED that units be CMIXF-12 compliant or among these three Unicode
characters.
Please take care not to confuse the accepted Unicode character
[`U+00B5`](https://codepoints.net/U+00B5) (µ) with the disallowed Unicode
character [`U+03BC`](https://codepoints.net/U+03BC) (μ):

> They look equal in appearance but are different to the computer.

Reviewer 2: Since U+00B5 maps to U+03BC but not the other way around, shouldn't U+03BC be the preferred value? That way data files which are modified by software using the compatibility mapping don't need to be manually fixed after they are modified.

Author: thanks @nellh that sounds reasonable - I have not heard of this mapping before -> do you have a reference where I could read up on this?

Reviewer 2: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize for JavaScript.

https://docs.python.org/3/library/unicodedata.html#unicodedata.normalize for Python.

Wikipedia also has a general summary under https://en.m.wikipedia.org/wiki/Unicode_equivalence

Author: > Since U+00B5 maps to U+03BC but not the other way around

Okay I get it now. Here is a Python example:

```Python
import unicodedata

# get utf-8 encoding of MICRO sign after NFC norm
unicodedata.normalize('NFC', 'µ').encode()                             
b'\xc2\xb5'

# same but NFKC norm --> MICRO sign is now SMALL GREEK MU
unicodedata.normalize('NFKC', 'µ').encode()                           
b'\xce\xbc'

# try same for SMALL GREEK MU: NFC norm doesn't change anything
unicodedata.normalize('NFKC', 'μ').encode()                           
b'\xce\xbc'

# also NFKC norm doesn't change anything
unicodedata.normalize('NFC', 'μ').encode()                            
b'\xce\xbc'
```

Units MUST consist of the `unit symbol` with an optionally accompanying
`prefix symbol` (see table below). Appropriate upper- or lower- casing MUST
be applied as declared by CMIXF-12.

For cases that are unspecified by this appendix, or the
[units section](../02-common-principles.md#units), the
[CMIXF-12](https://people.csail.mit.edu/jaffer/MIXF/CMIXF-12) convention
applies.
You can use the [cmixf Python package](https://github.com/sensein/cmixf) to

> check whether your formatting is compliant.

Reviewer 3: is there a corresponding issue on the BIDS-validator? Ideally, the validator should tell you instead of having to use another tool. Also works for people who are not familiar with Python.

Author: Not yet, but @rwblair is aware of it and I'll open an issue as soon as this PR is in to continue tracking this challenge.

Author: now there is an issue: https://github.com/bids-standard/bids-validator/issues/969

### Submultiples

| Prefix name                                 | Prefix symbol   | Factor           |
| ------------------------------------------- | --------------- | ---------------- |
| [deci](https://www.wikiwand.com/en/Deci-)   | d               | 10<sup>-1</sup>  |
| [centi](https://www.wikiwand.com/en/Centi-) | c               | 10<sup>-2</sup>  |
| [milli](https://www.wikiwand.com/en/Milli-) | m               | 10<sup>-3</sup>  |

> | [micro](https://www.wikiwand.com/en/Micro-) | u or µ (U+00B5) | 10<sup>-6</sup>  |

Reviewer 3: why don't we allow both micro and mu symbols? They are indistinguishable to the naked eye and it's mostly [a historical artifact](https://en.wikipedia.org/wiki/Micro-#Symbol_encoding_in_character_sets). I think we should make it easy  for people to adapt BIDS, not make them spend hours understanding the difference between these two symbols.

Author: > I think we should make it easy for people to adapt BIDS, not make them spend hours understanding the difference between these two symbols.

I agree. I guess we are balancing between narrowing down possibilities of formatting and being overly pedantic. In this case it's currently perhaps being overly pedantic 🙂 

@effigies brought this up in an online meeting not too long ago ... soo I'll update this soon, thanks @jasmainak

Reviewer: The reason that we didn't allow both was that it shifts more of the responsibility onto apps to parse the units. If we can say they're always MIXF, then all you have to know if MIXF. We had some suggestions for micro-, degree and Ohm (IIRC) that made this difficult by backwards compatibility, so we said MIXF + those three code points. We can just as easily say MIXF + mu + micro + degree + Ohm, but there will be a complexity cost to any software that cares about units.

Right now we're seeing a case where the validator requiring micro instead of mu is causing user confusion. We can relieve this momentarily by relaxing the constraint, but will we not simply be distributing the confusion to every tool's bug tracker?

Reviewer 3: humm ... I just went back and checked how we are dealing with this in MNE. We coerce the mu and micro into micro -- so it might be a tool-to-tool issue. But I agree that keeping one would be simpler for tool developers in the long run.

Reviewer 4: I'm with @jasmainak on this one -- the difference between these two characters is imperceptible, so I really don't like pushing this onto users.

I do understand your concern, though, @effigies. If we want to enforce this with the validator, then I think this case of mu vs micro needs a very specific message alerting users to exactly this difference. To my mind, that seems the only reasonable (humane ?) way to handle it.

What I don't know, though, is how easy this would be to accomplish with the current validator structure. WDYT ?

Author: I think it should be easily possible in the current validator structure:
- implement a new issue code Z to signal "you are using µ as a character, please be aware that ... do not use U+03BC ..." 

and then for validation:

- when parsing a `unit` column in a TSV file X
- upon encountering unicode `U+03BC` in a line Y of file X
- push an issue to issue list for file X at line Y: issue code Z

what's more of a challenge (and one of the reasons why this PR is stagnating) is to validate MIXF units as a whole :confused:

Author: re: validator -> @satra now has implemented a nice Python validator: https://github.com/sensein/cmixf

the JS version of a cmixf validator may follow after merging this PR, but it shouldn't be a hard blocker.

Reviewer 3: we should raise an issue and track these somewhere nonetheless

Author: agreed, I am now tracking our discussion here: https://github.com/bids-standard/bids-validator/issues/969

thanks for the many points.

Reviewer 2: Why not rely on Unicode normalization to solve the visually similar character issue? For example, the compatibility decomposition will map U+00B5 to U+03BC. Python and JavaScript both have support for these mappings in the standard libraries. bids-validator could just normalize all Unicode text before validation. The specification could also recommend which normalization is preferred for other software aiming to be compatible.

Author: @nellh thanks for your input and also the references (above) -> could you make a concrete proposal what needs to be changeds in this PR (using GitHub's code suggestion feature)?

Reviewer 5: I'd like to point out that according to the unicode standard ([PDF](https://www.unicode.org/reports/tr25/), p.11) the sign `U+03BC` is the preferred symbol. So it'd be surprising to some users to get a warning when using the preferred `U+03BC` instead of `U+00B5` (which according to unicode, should be recognized as well)

Author: Thanks @gpiantoni. I take your post as voting for option 3 in my summary (https://github.com/bids-standard/bids-specification/pull/411#issuecomment-633992688), right? ... what are your thoughts about option 1?

Reviewer 5: Thanks for asking my opinion. As end user, I'd prefer option 3. But I understand the complexity of implementing it, so option 1 is probably easier to implement (sorry, I don't know more than this to comment further). Option 2 however would put the BIDS guidelines in disagreement with the unicode standard.
